common/build/eslint-plugin-fluid/src/rules/no-unchecked-record-access.js

```diff
@@ -302,86 +301,150 @@ function isUndefinableIndexSignatureType(parserServices, node) {
-				if (right.arguments[0].type === "Identifier" && right.arguments[0].name === base) {
+				return true;
+			}
+		}
+
+		// Handle presence checks
+		if (parent.type === "IfStatement" || parent.type === "ConditionalExpression") {
+			if (parent.test) {
+				if (nodesAreEquivalent(parent.test, node)) {
 					return true;
 				}
+				if (parent.test.type === "BinaryExpression") {
+					if (parent.test.operator === "in") {
+						const testBase = getBaseObject(parent.test.right);
+						if (baseObj === testBase) {
+							// Check if the else block assigns the key
+							const ifStatement = parent;
+							const elseBlock = ifStatement.alternate;
+							if (elseBlock) {
+								const keyNode = parent.test.left;
+								if (
+									checkElseBlockAssignsKey(elseBlock, testBase, keyNode, context)
+								) {
+									return true;
+								}
+							}
+							return true;
+						}
+					}
+					if (
+						(parent.test.operator === "!==" || parent.test.operator === "!=") &&
+						((nodesAreEquivalent(parent.test.left, node) &&
+							isUndefinedNode(parent.test.right)) ||
+							(nodesAreEquivalent(parent.test.right, node) &&
+								isUndefinedNode(parent.test.left)))
+					) {
+						return true;
+					}
+				}
 			}
 		}
 
-		// Handle "in" operator checks
-		if (
-			(current.parent.type === "ConditionalExpression" ||
-				current.parent.type === "IfStatement") &&
-			current.parent.test?.type === "BinaryExpression" &&
-			current.parent.test.operator === "in"
-		) {
-			const test = current.parent.test;
-			const base = getBaseObject(node);
-			const testBase = getBaseObject(test.right);
-			if (base === testBase) {
-				return true;
+		// Additional check for 'in' checks in the same scope
+		const containingBlock = findContainingBlock(current);
+		if (containingBlock) {
+			for (const statement of containingBlock.body) {
+				if (statement.range[0] > current.range[0]) break; // Only check statements before current node
+				if (
+					statement.type === "IfStatement" &&
+					statement.test?.type === "BinaryExpression" &&
+					statement.test.operator === "in"
+				) {
+					const testBase = getBaseObject(statement.test.right);
+					const testKey = statement.test.left;
+					if (testBase === baseObj && nodesAreEquivalent(testKey, currentKeyNode)) {
+						const elseBlock = statement.alternate;
+						if (
+							elseBlock &&
+							checkElseBlockAssignsKey(elseBlock, testBase, testKey, context)
+						) {
+							return true;
+						}
+					}
+				}
 			}
 		}
 
-		current = current.parent;
+		current = parent;
 	}
 
-	if (
-		node.parent.type === "BinaryExpression" &&
-		node.parent.operator === "in" &&
-		node.parent.left === node
-	) {
-		return true;
+	return false;
+}
+
+function isUndefinedNode(node) {
+	return (
+		(node.type === "Identifier" && node.name === "undefined") ||
+		(node.type === "UnaryExpression" &&
+			node.operator === "void" &&
+			node.argument?.type === "Literal" &&
+			node.argument.value === 0)
```

Comment:
Nit: docs for this would be useful. I don't understand how this is equivalent to "undefined".

Reply:
This is for this weird edge case where undefined === void <expression>. I will add a doc
<img width="583" alt="image" src="https://github.com/user-attachments/assets/fe363c2d-a40d-49e9-8da1-35b484402594" />

Reply:
If `void <any expression>` returns `undefined`, should we remove the `node.argument.value === 0` check? Also, [the docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void#description) say that `void(0)` is equivalent to `void 0`, so do we need to check for the function-call version as well?